### PR TITLE
fix(types): replace 506 occurrences of 'Float64' with 'AbstractFloat'

### PR DESCRIPTION
**Pull request by bionicles**
This potentially breaks a bunch of stuff, but I'm keen to make a change like this because it would allow us to do probabilistic programming with neural networks... @andreasnoack and @matbesancon and others can you check this out? If we allow more flexibility with the Float type then we can more easily integrate with other libraries. I'm not familiar with this code base so please let's discuss what to look at to see if this is OK

Philosophically, it doesn't make sense to require extreme 64-bit precision numbers to talk about uncertainty, although I can understand why this could be required for numerical algorithms. Is there a safe way for us to convert between types of floats and preserve gradients etc?

If an algorithm did require 64bit numbers, and it converted 32bit inputs to 64bit, ran the procedure, and converted it back to 32bit, would this screw up Zygote and Chain Rules? (cc @MikeInnes and @oxinabox )

### CODE REVIEW

**Comment by bionicles** [comment]
Hmmm...VSCode seems to have screwed this up. I'll try again with Atom

**Comment by oxinabox** [comment]
Why stop at `AbstractFloat` ?
Seems better to go all the way to `Real` and then can use Rational number and fixed point numbers etc.

**Comment by oxinabox** [comment]
> If an algorithm did require 64bit numbers, and it converted 32bit inputs to 64bit, ran the procedure, and converted it back to 32bit, would this screw up Zygote and Chain Rules? (cc @MikeInnes and @oxinabox )

Its fine.

File: test/utils.jl [hunk 1]
```diff
@@ -41,7 +41,7 @@ N = GenericArray([1.0 0.0; 1.0 0.0])
 n = 10
 areal = randn(n,n)/2
 aimg  = randn(n,n)/2
-@testset "For A containing $eltya" for eltya in (Float32, Float64, ComplexF32, ComplexF64, Int)
+@testset "For A containing $eltya" for eltya in (Float32, AbstractFloat, ComplexF32, ComplexF64, Int)
```
Comment (johnczito) [inline test/utils.jl:44]: These changes are much too blunt, especially in the test files. Here (^) for example, where the point is to loop through different concrete `eltype`s.

Comment (oxinabox) [reply test/utils.jl:44]: Probably the tests should have been excluded from the regex.

**Comment by MikeInnes** [comment]
This is probably better done by hand (and perhaps one or two features at a time to make it easier to review). From a quick scan most of the changes are false positives or incorrect, e.g. changing the type parameters of output in examples or dispatch constraints like `Vector{AbstractFoat}` which aren't doing what you want.

Most likely significant work is needed to be truly generic (eg propagating the `eltype` of inputs) and changing some instances of `Float64` is a minority of it.

**Comment by oxinabox** [comment]
Ideally, SourceWalk.jl would be usable for this rough first pass rather than by hand, which would be a bit smarter, but still a manual pass would be needed after

**Comment by bionicles** [comment]
Ok, I swapped in the old test directory, next we can look at examples and 

Anyone know a better approach to get this done efficiently? I'm really not familiar with this codebase so it feels wrong for me to be tinkering here, but it would be super useful to hook distributions up to flux models

**Comment by bionicles** [comment]
I didn't even know Real type was a thing, I'll look at that. Guidance appreciated

**Comment by bionicles** [comment]
if we use Real, would that require us to convert Int to Float?

**Comment by bionicles** [comment]
hmmm, it looks like multiple dispatch freaks out. I thought AbstractFloat would accept Float64 but apparently a method defined for AbstractArray{AbstractFloat} doesn't match Array{Float64} ... super weird... are we really expected to define everything for every data type?

**Comment by bionicles** [comment]
AbstractMatrix{AbstractFloat} -> AbstractMatrix

**Comment by kmsquire** [comment]
> it looks like multiple dispatch freaks out. I thought `AbstractFloat` would accept `Float64` but apparently a method defined for `AbstractArray{AbstractFloat}` doesn't match `Array{Float64}` ... super weird... are we really expected to define everything for every data type? That would be a disgusting amount of busywork. Perhaps we should just delete the types alltogether

See https://docs.julialang.org/en/v1/manual/types/#Parametric-Composite-Types-1 for an explanation as to why this is true.

Instead of using `AbstractArray{AbstractFloat}`, you can use `AbstractArray{<:AbstractFloat}`.

**Comment by andreasnoack** [comment]
As mentioned in https://github.com/JuliaStats/Distributions.jl/pull/1126#issuecomment-640567003, this has to be done more carefully. E.g. some of these functions only work correctly for `Float64` inputs so I'm closing this one.

**Pull request closed without merging**